Projects/BookJSON2/main.go

```diff
@@ -20,89 +21,119 @@ type Author struct {
-	// if r.Method == POST
+func (b *BookHandler) book(w http.ResponseWriter, r *http.Request) {
 	switch r.Method {
-	case "POST": //Retriving a book with Id number
+	case http.MethodGet: //Retriving a book with Id number
 		{
-			w.Header().Set("content-type", "application/json")
-			jsonreqestfromweb := &Book{}
-			err := json.NewDecoder(r.Body).Decode(&jsonreqestfromweb)
-			if err != nil {
-				fmt.Println(err)
+
+			val := r.FormValue("id")
+			if val == "" {
+				if err := json.NewEncoder(w).Encode(b.Books); err != nil {
+					log.Println(err)
+					return
+				}
 				return
 			}
 
-			json.NewEncoder(w).Encode(b.Books[jsonreqestfromweb.Id])
-			break
+			if val == "break" {
+				w.WriteHeader(http.StatusBadRequest)
+				return
+			}
+
+			id, err := strconv.Atoi(val)
+			if err != nil {
+				panic(err)
+			}
+
+			defer r.Body.Close()
+			for i, value := range b.Books {
+				if value.Id == id {
+					if err := json.NewEncoder(w).Encode(b.Books[i]); err != nil {
+						log.Println(err)
+
+					}
+					return
+				}
+
+			}
+
 		}
-	case "PUT": // adding new book
-		{
 
-			w.Header().Set("content-type", "application/json")
+	case http.MethodPost: // adding new book
+		{
 			newbook := Book{}
 			_ = json.NewDecoder(r.Body).Decode(&newbook)
+			for _, item := range b.Books {
+				if newbook.Id == item.Id {
+					w.WriteHeader(http.StatusBadRequest)
+					return
+				}
+			}
 			b.Books = append(b.Books, newbook)
-			json.NewEncoder(w).Encode(newbook)
-			break
 
+			if err := json.NewEncoder(w).Encode(newbook); err != nil {
+				log.Println(err)
+			}
 		}
-	case "DELETE": //Delete a specific book data
+
+	case http.MethodDelete: //Delete a specific book data
 		{
 			w.Header().Set("content-type", "application/json")
 			jsonrequestfromweb := Book{}
 			_ = json.NewDecoder(r.Body).Decode(&jsonrequestfromweb)
 			b.Books = append(b.Books[:jsonrequestfromweb.Id], b.Books[jsonrequestfromweb.Id+1:]...)
-			json.NewEncoder(w).Encode(b.Books)
-			break
+			//json.NewEncoder(w).Encode(b.Books)
+
 		}
-	case "PATCH": //Update an existing book
+	case http.MethodPut: //Update an existing book
 		{
 			w.Header().Set("content-type", "application/json")
 			jsonrequestfromweb := Book{}
 			_ = json.NewDecoder(r.Body).Decode(&jsonrequestfromweb)
 			b.Books = append(b.Books[:jsonrequestfromweb.Id], b.Books[jsonrequestfromweb.Id+1:]...)
 			b.Books = append(b.Books, jsonrequestfromweb)
-			json.NewEncoder(w).Encode(b.Books)
-			break
+			//json.NewEncoder(w).Encode(b.Books)
+
 		}
 	}
 }
 
-func (b *Books) allbooks(w http.ResponseWriter, r *http.Request) {
+func (b *BookHandler) allbooks(w http.ResponseWriter, r *http.Request) {
 
 	switch r.Method {
 
-	case "GET": //fetch the list of all books
+	case http.MethodGet: //fetch the list of all books
 		{
-			w.Header().Set("content-type", "application/json")
-			json.NewEncoder(w).Encode(b.Books)
-			break
+			//w.Header().Set("content-type", "application/json")
```

Comment:
should not be commited

Reply:
Resolved with commit https://github.com/AmanSarraf/Go-Project/pull/9/commits/e5cc27e532ea86a404ba2b08ce15b5656cbe9ae2